\documentclass{csfyp}
\usepackage{graphicx}
\usepackage{enumitem} %for {enumerate}[nolistsep]
\usepackage{titlesec} %for titlespacing
\usepackage{pgfgantt} %for Gantt chart

\titlespacing*{\section}{0pt}{0.5\baselineskip}{\baselineskip}
\titlespacing*{\subsection}{0pt}{0.5\baselineskip}{\baselineskip}

\titleformat*{\section}{\Large\bfseries}
\titleformat*{\subsection}{\large\bfseries}

\title{The Malta Human Genome Project \\
  \large Progress Report}

\author{Sara Ann Abdilla (188396M)}
\supervisor{Jean Paul Ebejer}
\date{December 2016}

\begin{document}

\pagenumbering{roman} 
\tableofcontents
%\listoffigures
%\listoftables

%\maketitle
\newpage

\pagenumbering{arabic} 
\setcounter{page}{1}


\begin{abstract}
The aim of this report is to detail the groundwork accomplished on developing the visualisation tool and the genome browser required for the Malta Human Genome Project.  The problem is introduced with the motivation behind it along with the reasoning for its nontriviality.  The reader is then provided with the required background knowledge in order to properly discern the problem, followed by the project's objectives.  The current and intended techniques and evaluation strategy are explained.  Finally, the expected deliverables are listed with an accompanying Gantt chart.   
\end{abstract}


\section{Introduction \& Motivation}
\label{s:intro}

Over the years, more and more human reference genomes are being globally assembled.  A reference genome is a digital database consisting of DNA sequences; each individual's DNA corresponding to an arrangement of approximately 3 billion bases.  The possible nucleotide bases are adenine, cytosine, guanine and thymine - oftenly referred to by the letters A, C, G and T respectively \cite{aiBk, introgenom}.  

Genome assembly/sequencing technologies are rapidly advancing and their costs are decreasing; both factors being due to the fact that their importance is becoming more widely known.  DNA variations and mutations may correlate to diseases so discovering any approximately matching alignments between the reference genome and the DNA sequencing reads being analysed could very well help future medical diagnosis and treatments \cite{think}.

The University of Malta is developing a National Maltese Human Reference Genome for this reason whereby whole genome sequencing on certain Maltese DNA samples (provided by the Malta BioBank) will be performed.  The American human genome sequencing facility Complete Genomics, which was founded in 2006, will be a partner in this project. 

Thus, computation wise, a data visualisation tool along with a genome browser need to be constructed in order to aid in this endeavour.  


\section{Reasoning for Nontriviality of Problem}

Globally, large genome projects are being sequenced rapidly.  Examples of such projects include the 1000 Genomes Project and the International Cancer Genome Project \cite{bwtransform, refcompression, popgen}.  Malta should aim to be a part of this endevaour so that further studies can be conducted involving a larger variety of genes (i.e. the Maltese genes).  After all, while the reading of an individual's DNA shows the likeliness of that person developing a disease, the reading of a nationâ€™s DNA shows why that population is more likely to develop a disease \cite{think}.  Such analyses can therefore prove to not only medically aid global research but also nation-wide studies.             


\section{Background Research and Literature Review}

The development of a genome assembly technology consists of multiple stages; the main ones being genome compression, sequence alignment, alignment visualisation and genome browser construction; all the steps following one another.  The following points detail some research which has already been conducted in these areas.

\subsection{Genome Compression Tools}
Genomes, particularly human ones, consist of a large amount of data.  For this reason, in order to efficiently analyse them, said genomes are compressed using various methods.  For example, Chen et al devised the lossless {\textit{GenCompress}} algorithm which implements certain established compression algorithms such as Lempel-Ziv.  By analysing approximate matches based on the evaluated edit distances, it was found that such a method not only achieves the best compression ratio but also finds common sections in DNA sequences \cite{gencompress}.  Another example would be the lossless reference-based compression algorithm devised by Fritz et al.  It implements established components such as Golomb codes and De Bruijn graphs and was found to be quite efficient for read alignments similar to the reference genome \cite{refcompression}.  Other known algorithm are {\textit{Biocompress-2}} and {\textit{Cfact}} \cite{gencompress} along with {\textit{DNACompress}} and {\textit{ DNAZip}} \cite{refcompression}.          

\subsection{Read Alignment Tools}
Next-Generation Sequencing (NGS) technologies are evolving rapidly and to keep up with evolution, multiple read aligners are being produced; the most known being BWA, Bowtie, Soapv2, MAQ, BOAT, SHRiMP2 for the NGS platforms Illumina, Roche454 and ABI SOLiD.  The majority implement Burrows-Wheeler Transform (BWT) but there are also those which use FM-indexing, Smith-Waterman,  and Needleman-Wunsch algorithms \cite{bwtransform, popgen, cgreads}.  All these mentioned aligners are quite efficient relative to the task they are given; for example, while some aligners may be more efficient with short reads, others may be the opposite \cite{companalys, compalign}.  Other known algorithms which assist in approximate read mapping are the Hamming distance as well as the Levenshtein distance, both of which return a measure of the similarity between two sequences.

\subsection{Critique of Missing Studies}
Regarding alignment visualisation and genome browser construction, there do not seem to be any concrete algorithms for such cases.  This is a fundamental area in genome assembly technology which should be researched extensively; i.e. part of this project's goals.     

\section{Aims and Objectives}

The aim of this project is to build tools for the analysis of sequenced genomes from the Malta Human Genome Project.  The research areas studied are bioinformatics, big data, data storage, data visualisation and data analysis among others.

The main objectives of the system are as follows:

\begin{enumerate}[nolistsep]

   \item The compression of a reference genome such as that done by Chen et al. \cite{gencompress} and Fritz et al. \cite{refcompression} among others;

  \item The alignment of a reference genome against a number of sequencing reads such as that done by Lee et al. \cite{cgreads} and Li et al. \cite{bwtransform} among others;

  \item The data visualiation of human genomes;

  \item The construction of a genome browser using novel and established components in order to reference DNA for comparitive genomics and to analyse DNA mutations;  

  \item A comparitive review of existing methods against all the above points.

\end{enumerate}


\section{Methods and Techniques Used or Planned}

This section details the components which have been implemented along with a description of future plans.

\subsection{Genome Compression}
The genome compression method implemented makes use of certain in-built Python modules in order to compress the genome into a binary file by means of integers (each integer being of length 4 bytes).  Each nucleotide base is assigned two bits to represent it as there are only 4 possible bases. Chen et al. used the same deduction when developing {\textit{GenCompress}} \cite{gencompress}.  

\subsection{Sequence Alignment}
Four possible alignment methods were then taken into consideration - Hamming distance, Levenshtein/Edit distance, k-mer indexing and the most frequently used: FM-indexing with BWT.  These were first implemented and tested on string inputs; i.e. on an uncompressed genome of marginally less size than the human genome to confirm their applicability.  These same functions were then converted in order to instead support integer inputs due to the previously mentioned genome compression.  Analysing each algorithms' resulting match rate with the corresponding time taken led to the deduction that the ... algorithm performed best with ... matches done in ... seconds.     

\subsection{Alignment Visualisation}
The data visualisation tool's initial construction is based on {\textit{Tkinter}}, Python's standard Graphical User Interface (GUI) package.  The result is a depiction of the reads aligned with the genome, with lines repsresnting the position and length of each alignment/match.  As of now, its interactivity is in the form of clicking a line at a certain position to output the offset of the match at said point of clicking.     

\subsection{Planned Methods}
As for future development, firstly, the data visualisation tool needs to be further refined; i.e. become more interactive in terms of genome analysis.  Secondly, the genome browser tool also needs to be constructed using established technologies for efficient comparitive genomics.  Each method developed will be analysed and discussed to deduce the most feasible implementation.   


\section{Proposed Evaluation Stategy}

The main evaluation techniques proposed are as follows:

\begin{enumerate}[nolistsep]

  \item Comparitive review of the developed human genome visualisation tool with existing methods;

  \item Reference DNA and analyse DNA mutations using the constructed genome browser.

\end{enumerate}


\section{Expected Deliverables}

\begin{enumerate}[nolistsep]

  \item The Final Year Project (FYP) which will include: all the relevant background information required to understand said project, a detailed explanation of the system and the evaluation results; 

  \item The implementation of the designed and developed system;

  \item The documentation which will explain to the users how the system should be employed. 

The milestone schedule is represented in the following Gantt chart: \\

\noindent\resizebox{\textwidth}{!}{
\begin{ganttchart}[
    hgrid=true,
    vgrid=true,
    bar/.append style={fill=black!50},
]{1}{30}
\gantttitle{2016--2017}{30} \\
\gantttitle{Sep}{3} 
\gantttitle{Oct}{3} 
\gantttitle{Nov}{3} 
\gantttitle{Dec}{3} 
\gantttitle{Jan}{3} 
\gantttitle{Feb}{3} 
\gantttitle{Mar}{3} 
\gantttitle{Apr}{3} 
\gantttitle{May}{3}  
\gantttitle{Jun}{3}\\
\ganttbar{Research}{1}{12} \\
\ganttbar{Progress Report}{8}{11} \\
\ganttbar{Implementation}{4}{21} \\
\ganttbar{Dissertation}{21}{27} \\
\ganttbar{Review Report}{23}{27} \\
\ganttbar{Presentation}{27}{29} \\
\ganttbar{Exhibition}{29}{30}
\end{ganttchart}
}

\end{enumerate}


%\bibliographystyle{alpha}
\bibliographystyle{ieeetr}
\nocite{*}
\bibliography{references}

\end{document}
